dev/ToastNotifications/ToastNotificationManager.cpp

```diff
@@ -124,13 +131,49 @@ namespace winrt::Microsoft::Windows::ToastNotifications::implementation
         THROW_HR_IF(E_UNEXPECTED, notificationId == 0);
         toast.ToastId(notificationId);
     }
-    winrt::Windows::Foundation::IAsyncOperation<winrt::Microsoft::Windows::ToastNotifications::ToastProgressResult> ToastNotificationManager::UpdateToastProgressDataAsync(winrt::Microsoft::Windows::ToastNotifications::ToastProgressData /* data */, hstring /* tag */, hstring /* group */)
+
+    // API updates notification data when both tag and group are provided or when tag is provided but group is empty.
+    // For other scenarios the API returns ToastProgressResult::Failed.
+    winrt::Windows::Foundation::IAsyncOperation<winrt::Microsoft::Windows::ToastNotifications::ToastProgressResult> ToastNotificationManager::UpdateToastProgressDataAsync(winrt::Microsoft::Windows::ToastNotifications::ToastProgressData data, hstring tag, hstring group)
     {
-        throw hresult_not_implemented();
+        std::wstring appId{ RetrieveToastAppId() };
+
+        winrt::com_ptr<::ABI::Microsoft::Internal::ToastNotifications::IToastProgressData> toastProgressData
+            = winrt::make_self<ToastProgressDataABI>(data);
+
```

Comment:
I was thinking about this a bit more. If only the group is specified and the tag is nullptr or empty, I believe we should return ProgressResult::NotificationNotFound.

Exceptions are only for fatal errors. Not programmer errors and shouldn't be thrown under winrt APIs in principle.

Reply:
Is NotFound due to bad input by the developer, that they could and should have known better? Or because the underlying data changed (perhaps by another thread) and the developer couldn't have reasonably known and prevented?

i.e. is it like 'file not found' or 'array out of bounds'? The latter is something the developer can and should know and have prevented. The former can happen beyond the devs control so it's something they should be prepared to encounter

Reply:
Generally when we throw Notfound is: The developer tried to update notification data for a given combination of tag and group. when developer hasn't even posted a notification but tried to update notification data using the above API then we return Notification not found error. 

"I was thinking about this a bit more. If only the group is specified and the tag is nullptr or empty, I believe we should return ProgressResult::NotificationNotFound." -  I accept that this is not fatal error but I don't think NotificationNotfound is the right result to return. Maybe a different error code like ProgressResult::Unsupported?

Reply:
If we really look into the platform code though, even if there are multiple notifications with the same group, we cannot select between them. ie. we cannot find the right notification to replace. So NotFound sounds correct to me.

@DrusTheAxe  Yes this is a programmer error ie. By Design. No exceptions right on principle?

Reply:
Exceptions should only occur on excecptional conditions.

Bad developer input they could have prevented (e.g. array-out-of-bounds) is fair game for exceptions.

The canonical example I always think of is FileExists(). That has **3** possible outcomes:

1. File found == Success
2. File not found == Failure
3. Something went wrong == Error

'Something went wrong' can range from out-of-memory to disk subsystem threw an error to developer specified illegal input to the function e.g. filename=null when the function's documented as requiring a non-null filename.

Success+Failure should communicate their results w/o exceptions. Only Errors should throw exceptions e.g. FileExists() throwing an exception when not-found is a bad idea.

Reply:
ok.